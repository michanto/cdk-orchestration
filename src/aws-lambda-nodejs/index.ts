export * from './inline_nodejs_function';

