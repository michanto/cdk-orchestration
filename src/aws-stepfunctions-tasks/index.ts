export * from './late_bound_start_execution';

