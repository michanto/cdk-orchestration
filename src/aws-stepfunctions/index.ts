export * from './chainable';
export * from './states_transform';
export * from './insert_stepfunction_state';
