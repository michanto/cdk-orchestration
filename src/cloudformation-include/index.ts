export * from './cfninclude_to_cdk';
