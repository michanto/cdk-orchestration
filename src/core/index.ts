export * from './build_time';
export * from './post_resolve_token';
export * from './construct_tree_search';
export * from './construct_service';
export * from './construct_tree_service';
export * from './app_construct_tree_service';
export * from './stack_construct_tree_service';
export * from './construct_rtti';
export * from './construct_host';
export * from './logger';
export * from './log';
export * from './user_tokens';
export * from './singleton';
export * from './tree_inspectable';
export * from './service_inspector_aspect';
export * from './cfn_element_utils';
