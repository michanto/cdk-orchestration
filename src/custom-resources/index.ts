export * from './custom_resources_utilities';
export * from './encode_resource';
