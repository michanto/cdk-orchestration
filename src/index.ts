// Export constructs here
export * from './core';
export * as transforms from './transforms';
export * as aws_lambda_nodejs from './aws-lambda-nodejs';
export * as aws_stepfunctions_tasks from './aws-stepfunctions-tasks';
export * as cloudformation_include from './cloudformation-include';
export * as custom_resources from './custom-resources';
export * as orchestration from './orchestration';
