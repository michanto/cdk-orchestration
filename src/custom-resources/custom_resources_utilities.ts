import { CfnResource } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { BUILD_TIME, ConstructTreeSearch } from '../core';

/**
 * Utilities for creating custom resources
 */
export class CustomResourceUtilities {
  static isCustomResource(elt: Construct): elt is CfnResource {
    return CfnResource.isCfnResource(elt) && (elt as any)._cfnProperties.ServiceToken;
  }
  protected treeSearch = ConstructTreeSearch.for(CustomResourceUtilities.isCustomResource);

  /**
   * Returns the CfnResource that produces the custom resource.  This function throws
   * if there are none (or more than one).
   * @param target
   * @returns
   */
  findCustomResource(target: Construct) {
    let elements = this.treeSearch.searchDown(target);
    if (elements.length != 1) {
      throw new Error(`Construct ${
        target.node.path
      } is not a known custom resource type.  Found ${elements.length} CustomResources, expected one.`);
    }

    return elements.pop() as CfnResource;
  }

  /**
   * Always run a custom resource.  Throws if it cannot find one custom resource under target.
   * @param target - CustomResource, AwsCustomResource or similar.
   */
  runResourceAlways(target: Construct) {
    let resource = this.findCustomResource(target);

    resource.addPropertyOverride('Version', BUILD_TIME);
  }
}