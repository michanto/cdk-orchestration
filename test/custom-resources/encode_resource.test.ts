import { CfnCustomResource, CfnCustomResourceProps, Stack } from "aws-cdk-lib";
import { BadFunction } from "../util";
import { Template } from "aws-cdk-lib/assertions";
import { Construct } from "constructs";
import { EncodeResource } from "../../src/custom-resources";
import { CfnBucket } from "aws-cdk-lib/aws-s3";

describe('Custom Resource Utilities tests.', () => {
    it('EncodeResource works.', () => {
        // Given
        const stack = new Stack();
        let serviceToken = new BadFunction(stack, 'Fun').functionArn;
        new class extends CfnCustomResource{
            constructor(scope: Construct, id: string, props: CfnCustomResourceProps) {
                super(scope, id, props)
                this.addPropertyOverride("One", 1);
                this.addPropertyOverride("Two", true);
                this.addPropertyOverride("Three", "Value");
                EncodeResource.encodeCustomResource(this);
                // Encode twice to test double-encoding works.
                EncodeResource.encodeCustomResource(this, "Encode2");
            }
        }(stack, 'Res1', {
            serviceToken: serviceToken,
        });

        // THEN
        let template = Template.fromStack(stack).toJSON();
    
        expect(template).toMatchObject({
          Resources: {
            Res1: {
              Type: 'AWS::CloudFormation::CustomResource',
              Properties: {
                EncodedProperties: {
                    "Fn::Base64": JSON.stringify({
                        One: 1,
                        Two: true,
                        Three: "Value"
                    })
                },
                ServiceToken: expect.anything(),
              },
            },
          },
        });
    });

    it('EncodeResource not custom resource does nothing.', () => {
        // Given
        const stack = new Stack();
        let bucket = new CfnBucket(stack, "Bucket", { bucketName: "my_bucket"});
        bucket.addPropertyDeletionOverride
        new EncodeResource(bucket, "Encode")

        // THEN
        let template = Template.fromStack(stack).toJSON();
    
        expect(template).toMatchObject({
          Resources: {
            Bucket: {
                Properties: expect.not.objectContaining({
                    EncodedProperties: expect.anything(),
                    ServiceToken: expect.anything()
                })
            }
          },
        });
    });
});
