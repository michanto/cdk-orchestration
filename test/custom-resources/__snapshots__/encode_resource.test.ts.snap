// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Resource Utilities tests. EncodeResource no ServiceToken does nothing 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "FunA2CCED21": {
      "DependsOn": [
        "FunServiceRole3CC876D7",
      ],
      "Properties": {
        "Code": {
          "ZipFile": ""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = handler;
function log(message) {
    if (process.env.LogLevel) {
        console.log(JSON.stringify(message));
    }
}
async function handler(event, context) {
    log({ Event: event });
    log({ Context: context });
    return Promise.resolve(event ?? {});
}
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiZWNoby5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uLy4uLy4uLy4uL3NyYy9hd3MtbGFtYmRhLW5vZGVqcy9wcml2YXRlL3Rlc3RfbGFtYmRhcy9lY2hvLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBS0EsMEJBSUM7QUFURCxTQUFTLEdBQUcsQ0FBQyxPQUE0QjtJQUN2QyxJQUFJLE9BQU8sQ0FBQyxHQUFHLENBQUMsUUFBUSxFQUFFLENBQUM7UUFDekIsT0FBTyxDQUFDLEdBQUcsQ0FBQyxJQUFJLENBQUMsU0FBUyxDQUFDLE9BQU8sQ0FBQyxDQUFDLENBQUM7SUFDdkMsQ0FBQztBQUNILENBQUM7QUFDTSxLQUFLLFVBQVUsT0FBTyxDQUFDLEtBQVUsRUFBRSxPQUFZO0lBQ3BELEdBQUcsQ0FBQyxFQUFFLEtBQUssRUFBRSxLQUFLLEVBQUUsQ0FBQyxDQUFDO0lBQ3RCLEdBQUcsQ0FBQyxFQUFFLE9BQU8sRUFBRSxPQUFPLEVBQUUsQ0FBQyxDQUFDO0lBQzFCLE9BQU8sT0FBTyxDQUFDLE9BQU8sQ0FBQyxLQUFLLElBQUksRUFBRSxDQUFDLENBQUM7QUFDdEMsQ0FBQyIsInNvdXJjZXNDb250ZW50IjpbImZ1bmN0aW9uIGxvZyhtZXNzYWdlOiBSZWNvcmQ8c3RyaW5nLCBhbnk+KSB7XG4gIGlmIChwcm9jZXNzLmVudi5Mb2dMZXZlbCkge1xuICAgIGNvbnNvbGUubG9nKEpTT04uc3RyaW5naWZ5KG1lc3NhZ2UpKTtcbiAgfVxufVxuZXhwb3J0IGFzeW5jIGZ1bmN0aW9uIGhhbmRsZXIoZXZlbnQ6IGFueSwgY29udGV4dDogYW55KSB7XG4gIGxvZyh7IEV2ZW50OiBldmVudCB9KTtcbiAgbG9nKHsgQ29udGV4dDogY29udGV4dCB9KTtcbiAgcmV0dXJuIFByb21pc2UucmVzb2x2ZShldmVudCA/PyB7fSk7XG59XG4iXX0=",
        },
        "Handler": "handler",
        "Role": {
          "Fn::GetAtt": [
            "FunServiceRole3CC876D7",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunServiceRole3CC876D7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Res1": {
      "Properties": {
        "One": 1,
        "Three": "Value",
        "Two": true,
      },
      "Type": "AWS::CloudFormation::CustomResource",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
